// artus.v.2/fascisOperati/timer.js
'use strict';

const { artus } = require('artus.v.2');

const propObj = {
  preComutClosure: (closureForFn) => closureForFn,
  comutClosure: (closureForFn) => closureForFn,
  closureToMoresFn: (bfn, closureForFn) => (closureForFn, bfn),
  returnBfn: (bfn, closureForFn) => (closureForFn, bfn),
  methods: {},
  closureForFn: {
    context: {
      max: 0,
      counter: 0,
    },
    time: { msec: 0, },
    timer: {
      timeout: null,
      interval: null,
    },
    flags: {
      kill: false,
    }
  },
  thisForFn: {},
};



propObj.returnBfn = (bfn, closureForFn) => {
  if (!closureForFn.flags.kill) {
    return bfn;
  } else return () => undefined;
};

propObj.comutClosure = (closureForFn) => {
  if (closureForFn.context.counter === closureForFn.context.max) closureForFn.flags.kill = true;
  else closureForFn.context.counter++;
  return closureForFn;
};

propObj.methods = {
  printFn() {
    console.log(generativusFucus.propObj.closureForFn.bfn);
    return this;
  },
  limit(max) {
propObj.closureForFn.context.max = max;
    return this;
  },
  timer(msec) {
    const ms = propObj.closureForFn.time.msec = msec;
    const timer_ = setTimeout(() => {
      propObj.closureForFn.flags.kill = true;
      timer_.close();
    }, ms);
    return this;
  }
};


// propObj.closureForFn.context.max = 2;

// const fn = function(...args) {
//   console.dir({ args, this: this });
// };

// Use
// const fucus = generativusFucus(propObj, fn).limit(2).timer(200);
// fucus(0);
// setTimeout(() => {
//   fucus('primus');
// }, 100);
//
// setTimeout(() => {
//   fucus('secundus');
// }, 500);
debugger
const timer = artus.bind(null, propObj);

module.exports = { timer };
